bgpd/bgp_evpn.c

```diff
@@ -34,192 +39,2822 @@
+    {
+      if (tmp_vpn == vpn)
+        return 1;
+    }
+
+  return 0;
+}
+
+/*
+ * Compare Route Targets.
+ */
+static int
+evpn_route_target_cmp (struct ecommunity *ecom1, struct ecommunity *ecom2)
+{
+  if (ecom1 && !ecom2)
+    return -1;
+
+  if (!ecom1 && ecom2)
+    return 1;
+
+  if (!ecom1 && !ecom2)
+    return 0;
+
+  if (ecom1->str && !ecom2->str)
+    return -1;
+
+  if (!ecom1->str && ecom2->str)
+    return 1;
+
+  if (!ecom1->str && !ecom2->str)
+    return 0;
+
+  return strcmp(ecom1->str, ecom2->str);
+}
+
+/*
+ * Mask off global-admin field of specified extended community (RT),
+ * just retain the local-admin field.
+ */
+static inline void
+mask_ecom_global_admin (struct ecommunity_val *dst,
+                        struct ecommunity_val *src)
+{
+  u_char type;
+
+  type = src->val[0];
+  dst->val[0] = 0;
+  if (type == ECOMMUNITY_ENCODE_AS)
+    {
+      dst->val[2] = dst->val[3] = 0;
+    }
+  else if (type == ECOMMUNITY_ENCODE_AS4 ||
+           type == ECOMMUNITY_ENCODE_IP)
+    {
+      dst->val[2] = dst->val[3] = 0;
+      dst->val[4] = dst->val[5] = 0;
+    }
+}
+
+/*
+ * Map one RT to specified VNI.
+ */
+static void
+map_vni_to_rt (struct bgp *bgp, struct bgpevpn *vpn,
+               struct ecommunity_val *eval)
+{
+  struct irt_node *irt;
+  struct ecommunity_val eval_tmp;
+
+  /* If using "automatic" RT, we only care about the local-admin sub-field.
+   * This is to facilitate using VNI as the RT for EBGP peering too.
+   */
+  memcpy (&eval_tmp, eval, ECOMMUNITY_SIZE);
+  if (!is_import_rt_configured (vpn))
+    mask_ecom_global_admin (&eval_tmp, eval);
+
+  irt = lookup_import_rt (bgp, &eval_tmp);
+  if (irt && irt->vnis)
+    if (is_vni_present_in_irt_vnis (irt->vnis, vpn))
+      /* Already mapped. */
+      return;
+
+  if (!irt)
+    {
+      irt = import_rt_new (bgp, &eval_tmp);
+      assert (irt);
+    }
+
+  /* Add VNI to the hash list for this RT. */
+  listnode_add (irt->vnis, vpn);
+}
+
+/*
+ * Unmap specified VNI from specified RT. If there are no other
+ * VNIs for this RT, then the RT hash is deleted.
+ */
+static void
+unmap_vni_from_rt (struct bgp *bgp, struct bgpevpn *vpn,
+                   struct irt_node *irt)
+{
+  /* Delete VNI from hash list for this RT. */
+  listnode_delete (irt->vnis, vpn);
+  if (!listnode_head (irt->vnis))
+    {
+      list_free (irt->vnis);
+      import_rt_free (bgp, irt);
+    }
+}
+
+/*
+ * Create RT extended community automatically from passed information:
+ * of the form AS:VNI.
+ * NOTE: We use only the lower 16 bits of the AS. This is sufficient as
+ * the need is to get a RT value that will be unique across different
+ * VNIs but the same across routers (in the same AS) for a particular
+ * VNI.
+ */
+static void
+form_auto_rt (struct bgp *bgp, struct bgpevpn *vpn,
+              struct list *rtl)
+{
+  struct ecommunity_val eval;
+  struct ecommunity *ecomadd;
+
+  encode_route_target_as ((bgp->as & 0xFFFF), vpn->vni, &eval);
```

Comment:
Maybe it's easier to ignore this issue for this PR. I'll open a separate issue once this PR is merged to discuss that and maybe we could involve the editors of the draft to get their opinions. One of them is from Cisco, so he should know what Cisco plans to do about that.